listenbrainz/mbid_mapping_writer/mbid_mapper.py

```diff
@@ -0,0 +1,284 @@
+                return (hit, MATCH_TYPE_LOW_QUALITY)
+
+            # For exact matches, return exact match. duh.
+            if ac_dist == 0 and r_dist == 0:
+                self._log(Markup(f"""<b>exact_match</b>: ac_detuned {self.is_ac_detuned}, r_detuned {self.is_r_detuned}"""))
+                return (hit, MATCH_TYPE_EXACT_MATCH)
+
+            # If both fields are above the high quality threshold, call it high quality
+            if ac_dist <= self.MATCH_TYPE_HIGH_QUALITY_MAX_EDIT_DISTANCE and \
+               r_dist <= self.MATCH_TYPE_HIGH_QUALITY_MAX_EDIT_DISTANCE:
+                self._log(Markup(f"""<b>high quality</b>: ac_detuned {self.is_ac_detuned}, r_detuned {self.is_r_detuned},
+                                     ac_dist {ac_dist:.3f} r_dist {r_dist:.3f}"""))
+                return (hit, MATCH_TYPE_HIGH_QUALITY)
+
+            # If both fields are above the medium quality threshold, call it medium quality
+            if ac_dist <= self.MATCH_TYPE_MED_QUALITY_MAX_EDIT_DISTANCE and \
+                    r_dist <= self.MATCH_TYPE_MED_QUALITY_MAX_EDIT_DISTANCE:
+                self._log(Markup(f"""<b>med quality</b>: ac_detuned {self.is_ac_detuned}, r_detuned {self.is_r_detuned},
+                                     ac_dist {ac_dist:.3f} r_dist {r_dist:.3f}"""))
+                return (hit, MATCH_TYPE_MED_QUALITY)
+
+            # Poor results so far, lets try detuning fields and try again
+            if ac_dist > self.MATCH_TYPE_MED_QUALITY_MAX_EDIT_DISTANCE and ac_detuned:
+                ac_hit = ac_detuned
+                ac_detuned = ""
+                self.is_ac_detuned = True
+                self._log(Markup(f"""<b>no match</b>, ac_dist too high {ac_dist:.3f} continuing: detune ac"""))
+                continue
+
+            if r_dist > self.MATCH_TYPE_MED_QUALITY_MAX_EDIT_DISTANCE and r_detuned:
+                r_hit = r_detuned
+                r_detuned = ""
+                self.is_r_detuned = True
+                self._log(Markup(f"""<b>no match</b>, r_dist too high {r_dist:.3f} continuing: detune r"""))
+                continue
+
+            self._log(Markup(f"""<b>no good match</b>, ac_dist {ac_dist:.3f}, r_dist {r_dist:.3f}, moving on"""))
+            return (None, MATCH_TYPE_NO_MATCH)
+
+    def lookup(self, artist_credit_name_p, recording_name_p):
+
+        query = artist_credit_name_p + " " + recording_name_p
+        if self.remove_stop_words:
+            cleaned_query = []
+            for word in query.split(" "):
+                if word not in ENGLISH_STOP_WORD_INDEX:
+                    cleaned_query.append(word)
+
+            query = " ".join(cleaned_query)
+
+        search_parameters = {
+            'q': query,
+            'query_by': "combined",
+            'prefix': 'no',
+            'num_typos': self.MATCH_TYPE_MED_QUALITY_MAX_EDIT_DISTANCE
+        }
+
+        while True:
+            try:
+                hits = self.client.collections[COLLECTION_NAME].documents.search(
```

Comment:
Despite all the fixes, one recording I mentioned earlier never got mapped. Looking at the [explain](https://labs.api.listenbrainz.org/explain-mbid-mapping?artist_credit_name=Selena+Gomez%2C+Rauw+Alejandro&recording_name=Baila+Conmigo+%28with+Rauw+Alejandro%29), I saw `HIT: artist: selena gomez w rauw alejandro recording: baila conmigo`.  This is interesting because of the `w` in there. Orginally, I thought it might be a with being clipped down to w somehow turns out its not.

I manually queried typesense for this query to find the recording mbid, it was hitting. It is [f90720d0-3c4e-42a9-94f1-df26dd4ad472](https://musicbrainz.org/recording/f90720d0-3c4e-42a9-94f1-df26dd4ad472).

The artist credit name on this recording is `Selena Gomez w/ Rauw Alejandro`. We remove the punctuation so `w/` becomes `w` and we get what we see above. Now looking at that recording's page, I do see the with version there but its the release artist name or track artist name. So I can see how this recording did not get matched. I don't know enough about this stuff to tell whether track_artist and release_artist fields are worthwhile in considering matches but it looks like a possible way to improve matches.

Looking closely at the explains again, we see it tried original first, the ac_dist was within limits but not r_dist. It then detunes artist credit which actually increases the ac_dist and then tries a detuned recording name but by this time ac_dist has become bigger than the threshold so no match. But if we had detuned recording first instead of artist credit, we would have got a match. So now, I think it makes sense to add the case for `original ac r detuned ` as suggested in the other comment. 

(Alternatively, try to detune on the field which has a greater dist first. But I think covering all cases if probably better.)